java/core/src/main/java/com/microsoft/bond/BondType.java

```diff
@@ -0,0 +1,514 @@
+    private static final ConcurrentHashMap<
+            Class<? extends BondEnum>,
+            EnumBondType<? extends BondEnum>> enumTypeRegistry =
+            new ConcurrentHashMap<
+                    Class<? extends BondEnum>,
+                    EnumBondType<? extends BondEnum>>();
+
+    // precomputed hash code helps to speed up object retrieval from the type cache
+    private final int precomputedHashCode;
+
+    // restrict subclasses to the current package
+    BondType(int precomputedHashCode) {
+        this.precomputedHashCode = precomputedHashCode;
+    }
+
+    /**
+     * Gets the simple name of the Bond type, which excludes the namespace.
+     * The names of generic type parameters, if any, are not included in the name.
+     *
+     * @return short simple type name
+     */
+    public abstract String getName();
+
+    /**
+     * Gets the qualified name of the Bond type, which includes the namespace.
+     * The names of generic type parameters, if any, are not included in the name.
+     *
+     * @return short namespace-qualified type name
+     */
+    public abstract String getQualifiedName();
+
+    /**
+     * Gets the full name of the Bond type, which includes namespaces and generic type parameters.
+     * This method is used in exception reporting.
+     *
+     * @return full type name
+     */
+    public final String getFullName() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(this.getQualifiedName());
+        BondType<?>[] typeArguments = this.getGenericTypeArguments();
+        if (typeArguments != null) {
+            sb.append("<");
+            sb.append(typeArguments[0].getFullName());
+            for (int i = 1; i < typeArguments.length; ++i) {
+                sb.append(", ");
+                sb.append(typeArguments[i].getFullName());
+            }
+            sb.append(">");
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Gets the {@link BondDataType} value for this type.
+     *
+     * @return the bond data type enumeration value
+     */
+    public abstract BondDataType getBondDataType();
+
+    /**
+     * Gets the {@link Class} instance for the values (objects) of this type.
+     * The actual class of a value of this Bond type may not be the same as the returned class,
+     * but its is always assignable to it (i.e. a subclass or a Java primitive type).
+     *
+     * @return the class object
+     */
+    public abstract Class<T> getValueClass();
+
+    /**
+     * Gets the {@link Class} instance for the Java primitive values (non-objects) of this type
+     * or null if a Java primitive type does not exist. This method complements the
+     * {@link BondType#getValueClass()} method which returns the classes for object instances.
+     *
+     * @return the class object for a primitive type or null if a primitive type doesn't exists
+     */
+    public abstract Class<T> getPrimitiveValueClass();
```

Comment:
Added comments.

---
In reply to: [123091392](https://github.com/Microsoft/bond/pull/502#discussion_r123091392) [](ancestors = 123091392)